### ISSUE TITLE
GameServerQuery helps attacker get server offline.

### ISSUE THREAD

**Issue by iNilo** [GameServerManagers/LinuxGSM#473]
Hey, Im opening this issue to make you aware of an rather unfortunate issue.

I'm hosting some CSGO Servers and I'm sure it's purely accidental that this happened but when an attacker UDP floods the srcds ports (27015) the srcds will stop responding on those ports (out of self protection?).

So when running a cronjob every minute that checks for an offline server the 5 millisecond timeout is causing a restart.

https://github.com/dgibbs64/linuxgsm/blob/master/GameServerQuery/gsquery.py#L17

TLDR: attacker floods ports, query script thinks server is offline, attacker succeeds in taking down a (full) server.

This might be an unsolvable problem on your part and may require the server owners to protect against UDP floods, but I want to let you know this is happening

**Comment by dgibbs64** [comment]
Thanks for this info Will have to see if there is a way around this problem. Maybe there is some sort of whitelist for SRCDS or something. Will have to play around. However if anyone does find a way around this issue please post it.

**Comment by iNilo** [comment]
My first idea is to RCON query the server since that is TCP and afaik the RCON was still available when under attack

**Comment by dgibbs64** [comment]
That may be possible. Will take a while to implement something though

**Comment by pereirtc** [comment]
Hi,

this can be solved with the following iptables rule:

-A INPUT -p udp -m udp --dport 27015 -m state --state NEW -m recent --set --name DEFAULT --rsource
-A INPUT -p udp -m udp --dport 27015 -m state --state NEW -m recent --update --seconds 60 --hitcount 10  --name DEFAULT --rsource -j DROP


basically this rule will drop anyone who send more than 10 requests in less than 60 seconds. those numbers are not mandatory and you can adapt for your needs.

**Comment by dgibbs64** [comment]
Thats a really useful bit of info

**Comment by iNilo** [comment]
Yea, not just useful its awesome. just reminds me how powerful iptables are and should really try and understand the iptables chains

**Comment by digimon-dk** [comment]
I'm new to iptables . is it right to do it like this in ubuntu ?? 
command:
sudo iptables -A INPUT -p udp -m udp --dport 27015 -m state --state NEW -m recent --set --name DEFAULT --rsource
-A INPUT -p udp -m udp --dport 27015 -m state --state NEW -m recent --update --seconds 60 --hitcount 10 --name DEFAULT --rsource -j DROP

Thanks for a awesome work :+1:

**Comment by dgibbs64** [comment]
The UFW is probably a better option and simpler  for Ubuntu. My understanding is that is uses iptables but makes it simple to use

https://wiki.ubuntu.com/UncomplicatedFirewall

Here is a tutorial that may help
https://www.digitalocean.com/community/tutorials/how-to-set-up-a-firewall-with-ufw-on-ubuntu-14-04

**Comment by AlexCSilva** [comment]
Nice info on the iptables etc, I think I had similar problem on my Gmod server haha

**Comment by dgibbs64** [comment]
This could do with updating on the WIKI. I require a knowledgeable person to create a brief page about firewalls for Ubuntu and CentOS and how to protect your query port

**Comment by UltimateByte** [comment]
Will do it.

**Comment by UltimateByte** [comment]
Is this enough for now ?
https://github.com/GameServerManagers/LinuxGSM/wiki/Firewalls

**Comment by UltimateByte** [comment]
Well, i guess it's enough. :)
